### PR TITLE
Annotation System

### PR DESCRIPTION
**Pull request by CAG2Mark**
Adds an annotation system and basic type checker support (TODO). 

The parser supports adding arbitrary annotations to any expressions/statements. For example:
```fs
@tailrec
fun sum(n) = if n == 0 then 0 else @tailrec sum(n - 1)
```
This is in preparation for implementing the `@tailrec` annotation.

TODO:
- Currently it uses a mutable style `Annotated` trait similar to `Located`. Change it to use an `Ann` syntax that contains other terms and contains a list of annotations.
- Implement pretty-printing for the changes.
- Basic type checking to make sure all annotation names are in scope.

### CODE REVIEW

**Comment by CAG2Mark** [comment]
Added an `Ann` case class that extends `Term` and modified the parser to parse to this.

TODO: 
- Figure out the best way to store annotations in non-Term statements that we still want to annotate such as `NuFunDef`.  Possible solutions: 
  - Just add an `annotations` field to each of these.
  - Create an `AnnStatement` class, though this is quite messy and also forces us to also consider the case when `AnnStatement` contains a `Term`, which feels unclean.
- (*need help*) Missing case for `Ann` in `ClassLifter` class `liftTerm`. Still need to figure out what this function does.
- (*need help*) In `compiler/Helpers:32`, need to figure out whether this is a conversion we actually want, i.e. whether we should preserve annotations when converting to `Expr`. If I'm understanding correctly, `Expr` represents nodes in a control flow graph, so we probably don't need to preserve annotations (?)
- Update the type checker.

**Comment by LPTK** [comment]
> Figure out the best way to store annotations in non-Term statements that we still want to annotate such as `NuFunDef`. Possible solutions:
> * Just add an `annotations` field to each of these.

Yes, this is the way. In practice that means you only have to add the field to `NuFunDef` and `NuTypeDef`. Terms already support annotations through `Ann`, and other statements are either deprecated or won't support annotations.






> * (_need help_) Missing case for `Ann` in `ClassLifter` class `liftTerm`. Still need to figure out what this function does.

Just add it to the catch-all case of unsupported stuff, for now. This is experimental code that's already being changed anyway.

> * (_need help_) In `compiler/Helpers:32`, need to figure out whether this is a conversion we actually want, i.e. whether we should preserve annotations when converting to `Expr`. If I'm understanding correctly, `Expr` represents nodes in a control flow graph, so we probably don't need to preserve annotations (?)

Sam remark. This code is already obsolete.

> * Update the type checker.

Easy. Just defer the type checking of `Ann(Var(nme) :: anns, receiver)` to that of `Var(nme)` and `Ann(anns, receiver)`.

BTW I'm not convinced we need multiple annotations in the same `Ann` term. This has the downside of allowing the dummy `Ann(Nil, t)` and it's also kind of unnecessary. Class `Term` could be given a helper method `peelAnnotations` that would return a `(List[Var], Term)`, if that's even needed.

File: compiler/shared/main/scala/mlscript/compiler/Helpers.scala [hunk 1]
```diff
@@ -29,6 +29,7 @@ object Helpers:
 
   def term2Expr(term: Term): Expr = {
       term match
+        case Ann(anns, receiver) => term2Expr(receiver) // TODO: Make sure this is actually what we want.
```
Comment (LPTK) [inline compiler/shared/main/scala/mlscript/compiler/Helpers.scala]: Don't sweat it ;^)
```suggestion
        case Ann(anns, receiver) => term2Expr(receiver)
```

File: shared/src/main/scala/mlscript/NewParser.scala [hunk 2]
```diff
@@ -8,6 +8,8 @@ import utils._, shorthands._
 import mlscript.Message._
 import BracketKind._
 
+import mlscript.Annotation
+
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NewParser.scala]: This import does nothing.
```suggestion
```

File: shared/src/main/scala/mlscript/Typer.scala [hunk 3]
```diff
@@ -817,6 +817,9 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
       case v @ Var("_") =>
         if (ctx.inPattern || funkyTuples) freshVar(tp(v.toLoc, "wildcard"), N)
         else err(msg"Widlcard in expression position.", v.toLoc)
+      
+      // TODO: Check all annotations are in scope
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/Typer.scala]: Don't forget to handle that.

File: shared/src/main/scala/mlscript/helpers.scala [hunk 4]
```diff
@@ -985,6 +987,7 @@ trait StatementImpl extends Located { self: Statement =>
   }
   
   def children: List[Located] = this match {
+    case Ann(_, trm) => trm :: Nil
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/helpers.scala]: The annotations should be part of the children!

File: shared/src/main/scala/mlscript/syntax.scala [hunk 5]
```diff
@@ -47,6 +47,14 @@ final case class MethodDef[RHS <: Term \/ Type](
   val children: Ls[Located] = nme :: body :: Nil
 }
 
+// Annotations
+final case class Annotation(name: Var) extends Located {
+  def children: List[Located] = name :: Nil
+}
+
+// final case class Ann(anns: Ls[Annotation], receiver: Statement) extends Statement
+// final case class AnnDesugared(anns: Ls[Annotation], receiver: DesugaredStatement) extends DesugaredStatement
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/syntax.scala]: Don't forget to clean up.

File: shared/src/test/diff/nu/Annotations.mls [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+//│ Parsed: {fun hello = (x,) => {let rec x = @hello 2}}
+//│ AST: TypingUnit(List(NuFunDef(None,Var(hello),None,List(),Left(Lam(Tup(List((None,Fld(_,Var(x))))),Blk(List(NuFunDef(Some(true),Var(x),None,List(),Left(Ann(List(Annotation(Var(hello))),IntLit(2)))))))))))
+//│ Parsed: 
+//│ 
+
+
+:p
+fun x =
+    let y = 2
+    2
```
Comment (LPTK) [inline shared/src/test/diff/nu/Annotations.mls]: Test case seems redundant?

File: shared/src/test/diff/nu/Annotations.mls [hunk 7]
```diff
@@ -0,0 +1,50 @@
+:NewDefs
```
Comment (LPTK) [inline shared/src/test/diff/nu/Annotations.mls:1]: Also test annotations on classes, modules, and vals.

**Comment by CAG2Mark** [comment]
> BTW I'm not convinced we need multiple annotations in the same Ann term. This has the downside of allowing the dummy Ann(Nil, t) and it's also kind of unnecessary.

True, we can just store it as `Ann(annotation1, Ann(annotation2, ...))`.

**Comment by CAG2Mark** [comment]
It seems that adding an `annotations` field to `NuTypeDef` or `NuFunDef` creates errors in 70+ other places. Is there a better way of doing this?

**Comment by LPTK** [comment]
For such secondary fields, we normally add them to the second parameter list, which creates fewer errors as it does not affect the pattern syntax. Just add another `val` in there. You'll still need to initialize and propagate a bunch of them, but these are not very hard changes.

**Comment by CAG2Mark** [comment]
TODO: Fix codegen -- probably ok to just discard annotations for now. Maybe throw an error if `@tailrec` is seen, since it is not yet implemented.

File: shared/src/test/diff/nu/UserDefinedAnnotations.mls [hunk 8]
```diff
@@ -14,9 +14,48 @@ fun foo = 0
 class tailrec2() extends Annotation
 //│ class tailrec2() extends Annotation
 
-// :e // TODO should be rejected
+:e // In the future, this should be supported.
```
Comment (LPTK) [inline shared/src/test/diff/nu/UserDefinedAnnotations.mls]: Why should this be supported? The code is wrong.

File: shared/src/main/scala/mlscript/NewParser.scala [hunk 9]
```diff
@@ -621,8 +653,16 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
     raise(WarningReport(msg"[${cur.headOption.map(_._1).mkString}] ${""+msg}" -> loco :: Nil,
       newDefs = true))
   
-  final def exprOrIf(prec: Int, allowSpace: Bool = true)(implicit et: ExpectThen, fe: FoundErr, l: Line): IfBody \/ Term = wrap(prec, allowSpace) { l =>
-    cur match {
+  final def exprOrIf(prec: Int, allowSpace: Bool = true, annotations: Ls[Annotation] = Nil)(implicit et: ExpectThen, fe: FoundErr, l: Line): IfBody \/ Term = wrap(prec, allowSpace) { l =>
+    val moreAnnotations: Ls[Annotation] = parseAnnotations(false)
+
+    if (!moreAnnotations.isEmpty) {
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NewParser.scala]:
```suggestion
    if (moreAnnotations.nonEmpty) {
```

File: shared/src/main/scala/mlscript/NewParser.scala [hunk 10]
```diff
@@ -851,7 +891,24 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
...
+    }
+    
+    if (allAnns.isEmpty) res
+    else res match {
+      case Left(body) => body match {
+        case IfThen(expr, rhs) =>
+          Left(IfThen(wrapAnns(expr, allAnns), rhs))
+        case _ =>
+          err(msg"Unexpected annotation" -> allAnns.head.toLoc :: Nil)
+          L(body) // discord annotations for now
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NewParser.scala]:
```suggestion
          L(body) // discard annotations for now
```

File: shared/src/main/scala/mlscript/NuTypeDefs.scala [hunk 11]
```diff
@@ -1127,6 +1127,13 @@ class NuTypeDefs extends ConstraintSolver { self: Typer =>
                 if (fd.tparams.nonEmpty)
                   err(msg"Type parameters are not yet supported in this position",
                     fd.tparams.head.toLoc)
+              
+              implicit val gl: GenLambdas =
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NuTypeDefs.scala]: Let's not do that. The comment refers to a very specific call to `typeTerm`. There's no reason to type annotations with the same value of `GenLambdas`. For annotations, just use `false`.

File: shared/src/main/scala/mlscript/NuTypeDefs.scala [hunk 12]
```diff
@@ -1404,6 +1413,13 @@ class NuTypeDefs extends ConstraintSolver { self: Typer =>
                   case _ => die
                 }
               }
+
+              // Check annotations
+              td.annotations.foreach(ann => {
+                implicit val gl: GenLambdas = false // TODO: what should this be?
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NuTypeDefs.scala]: `false` is fine. It's irrelevant here.

File: shared/src/main/scala/mlscript/NuTypeDefs.scala [hunk 13]
```diff
@@ -1404,6 +1413,13 @@ class NuTypeDefs extends ConstraintSolver { self: Typer =>
                   case _ => die
                 }
               }
+
+              // Check annotations
+              td.annotations.foreach(ann => {
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NuTypeDefs.scala]:
```suggestion
              td.annotations.foreach { ann =>
```

File: shared/src/main/scala/mlscript/Typer.scala (hunk 3 above)
Comment (LPTK) [reply shared/src/main/scala/mlscript/Typer.scala]: My remark was also meant to remind you to remove the comment! Unresolved.

File: shared/src/main/scala/mlscript/Typer.scala [hunk 14]
```diff
@@ -1736,7 +1746,7 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
       case TypedNuAls(level, td, tparams, body) =>
         ectx(tparams) |> { implicit ectx =>
           NuTypeDef(td.kind, td.nme, td.tparams, N, N, S(go(body)), Nil, N, N, TypingUnit(Nil))(
-            td.declareLoc, td.abstractLoc)
+            td.declareLoc, td.abstractLoc, Nil)
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/Typer.scala]: Why not `td.annotations`?

File: shared/src/main/scala/mlscript/helpers.scala [hunk 15]
```diff
@@ -531,6 +531,7 @@ trait TermImpl extends StatementImpl { self: Term =>
   def describe: Str = sugaredTerm match {
     case S(t) => t.describe
     case N => this match {
+      case Ann(annotations, receiver) => receiver.describe
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/helpers.scala]:
```suggestion
      case Ann(_, Ann(_, receiver)) => receiver.describe
      case Ann(_, receiver) => "annotated " + receiver.describe
```

File: shared/src/main/scala/mlscript/helpers.scala [hunk 16]
```diff
@@ -985,6 +987,10 @@ trait StatementImpl extends Located { self: Statement =>
   }
   
   def children: List[Located] = this match {
+    case Ann(ann, trm) => trm match {
+      case t: Ann => ann :: t.children
+      case t => ann :: t :: Nil
+    }
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/helpers.scala]: Strange idea. Why not just `case Ann(ann, trm) => ann :: trm :: Nil`?

File: shared/src/main/scala/mlscript/syntax.scala [hunk 17]
```diff
@@ -47,6 +47,11 @@ final case class MethodDef[RHS <: Term \/ Type](
   val children: Ls[Located] = nme :: body :: Nil
 }
 
+// Annotations
+final case class Annotation(name: Var) extends Located {
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/syntax.scala]: Didn't we conclude that this class is useless? Just use a term in `Ann`.

File: shared/src/main/scala/mlscript/syntax.scala [hunk 18]
```diff
@@ -216,14 +222,15 @@ final case class NuFunDef(
   nme: Var,
   symbolicNme: Opt[Var],
   tparams: Ls[TypeName],
-  rhs: Term \/ Type,
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/syntax.scala]: Please don't...

File: shared/src/test/diff/nu/UserDefinedAnnotations.mls (hunk 8 above)
Comment (CAG2Mark) [reply shared/src/test/diff/nu/UserDefinedAnnotations.mls]: My bad, I should've stated it more clearly. I meant that we should add support for annotations with parameters in the future.

File: shared/src/main/scala/mlscript/helpers.scala (hunk 16 above)
Comment (CAG2Mark) [reply shared/src/main/scala/mlscript/helpers.scala]: Ah, I tried to keep the semantics of `Ann(Ls[Annotation], term)` and have all the annotations in one list. This isn't necessary with the updated definition of `Ann`, so I'll change this.

File: shared/src/test/diff/nu/UserDefinedAnnotations.mls (hunk 8 above)
Comment (LPTK) [reply shared/src/test/diff/nu/UserDefinedAnnotations.mls]: Got it. Then why not add a new test case using `@tailrec2()` and move the comment to that one?

Comment (CAG2Mark) [reply shared/src/test/diff/nu/UserDefinedAnnotations.mls]: Perhaps I should just leave it in a comment, and not as an actual test. Such a test case would actually pass, because it will add a `tailrec2` annotation to the unit literal `()`, which feels wrong. I'll need to make some non-trivial changes to the parser to implement this in the end.

EDIT: Seems it doesn't pass, so I'll add it as a proper test

File: shared/src/test/diff/nu/UserDefinedAnnotations.mls [hunk 19]
```diff
@@ -59,3 +59,16 @@ module dummy
 //│ ║  l.56: 	@helloa
 //│ ╙──      	 ^^^^^^
 //│ module dummy
+
+// TODO: In the future, we should still support annotations that require a constructor.
+:e
```
Comment (LPTK) [inline shared/src/test/diff/nu/UserDefinedAnnotations.mls]: Comments starting with "TODO" are interpreted specially (a wart we should remove).
```suggestion
:e // TODO: In the future, we should still support annotations that require a constructor.
```

**Review by LPTK — APPROVED** [review]
Otherwise it looks good to me. Nice feature!

**Comment by CAG2Mark** [comment]
> TODO: Fix codegen -- probably ok to just discard annotations for now. Maybe throw an error if `@tailrec` is seen, since it is not yet implemented.

I see you've approved the changes, but I suppose this should be done before merging into master?

File: shared/src/test/diff/nu/UserDefinedAnnotations.mls (hunk 19 above)
Comment (CAG2Mark) [reply shared/src/test/diff/nu/UserDefinedAnnotations.mls]: In this case, is keeping it as a TODO fine?

**Comment by LPTK** [comment]
Right. Forgot about that one, but it should be trivial.

I any case, I'll do the merge when it's ready.

**Comment by CAG2Mark** [comment]
My bad, probably should have rebased instead of merging. Anyway, it should be ready to merge now.

File: shared/src/test/diff/nu/Annotations.mls [hunk 20]
```diff
@@ -39,7 +48,7 @@ fun foo(n) =
   else ()
 //│ fun foo: Int -> ()
 
-// TODO
+:NoJS // recurses infinitely since `div` is float division. do not run it for now
```
Comment (LPTK) [inline shared/src/test/diff/nu/Annotations.mls]: Why not just increase the threshold to `n > 0.5` instead of writing a full comment and disabling JS? :P

File: shared/src/main/scala/mlscript/JSBackend.scala [hunk 21]
```diff
@@ -76,9 +76,9 @@ abstract class JSBackend(allowUnresolvedSymbols: Bool) {
       translatePattern(base)
     case Inst(bod) => translatePattern(bod)
     case Ann(ann, receiver) => translatePattern(receiver)
-    case _: Lam | _: App | _: Sel | _: Let | _: Blk | _: Bind | _: Test | _: With | _: CaseOf
-        | _: Subs | _: Assign | _: If | _: New  | NuNew(_) | _: Splc | _: Forall | _: Where
-        | _: Super | _: Eqn | _: AdtMatchWith | _: Rft | _: While =>
+    case _: Lam | _: App | _: Sel | _: Let | _: Blk | _: Bind | _: Test | _: With | _: CaseOf | _: Subs | _: Assign
+        | If(_, _) | New(_, _)  | NuNew(_) | _: Splc | _: Forall | _: Where | _: Super | _: Eqn | _: AdtMatchWith
+        | _: Rft | _: While | _: Quoted | _: Unquoted =>
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/JSBackend.scala]: What happened here? Why all these unnecessary changes?

Note: I always prefer writing things like `_: If` instead of `If(_, _)` as the former will need to be adapted if/when `If` gets a new argument.

Comment (LPTK) [reply shared/src/main/scala/mlscript/JSBackend.scala]: Never mind I just realize this is probably because of the merge.

**Comment by LPTK** [comment]
> My bad, probably should have rebased instead of merging. Anyway, it should be ready to merge now.

No, I know people rebase branches like this, but it's an anti-pattern. It is likely to make the commits in the middle broken, causes unnecessary conflicts, and is annoying for people (like me) who've already checked out your branch. I'm going to squash the PR on merge anyway, as its history isn't interesting/useful.

**Comment by CAG2Mark** [comment]
> > My bad, probably should have rebased instead of merging. Anyway, it should be ready to merge now.
> 
> No, I know people rebase branches like this, but it's an anti-pattern. It is likely to make the commits in the middle broken, causes unnecessary conflicts, and is annoying for people (like me) who've already checked out your branch. I'm going to squash the PR on merge anyway, as its history isn't interesting/useful.

Got it. What's the usual procedure for merging into the main branch here?

**Comment by LPTK** [comment]
It really depends. For short-lived PRs like this with many almost meaningless commits (like applying review suggestions), I prefer to squash. But if the branch has a lot of history by several authors or/and has been used as a base of other branches, I use a merge commit, to retain history details and reproducible commits (ie which compile and whose tests are usually green).
